server/server.go

```diff
@@ -845,6 +845,53 @@ func (j *jsonRPCHub) TraceBlock(
 	return results, nil
 }
 
+// IntermediateRoots executes a block (bad- or canon- or side-), and returns a list
+// of intermediate roots: the stateroot after each transaction.
+func (j *jsonRPCHub) IntermediateRoots(
+	block *types.Block,
+	tracer tracer.Tracer,
+) ([]types.Hash, error) {
+	if block.Number() == 0 {
+		return nil, errors.New("genesis block can't have transaction")
+	}
+
+	parentHeader, ok := j.GetHeaderByHash(block.ParentHash())
+	if !ok {
+		return nil, errors.New("parent header not found")
+	}
+
+	blockCreator, err := j.GetConsensus().GetBlockCreator(block.Header)
+	if err != nil {
+		return nil, err
+	}
+
+	transition, err := j.BeginTxn(parentHeader.StateRoot, block.Header, blockCreator)
+	if err != nil {
```

Comment:
I would suggest this simplification (also probably we can remove the `GetBlockCreator`, since IMO it does not do anything useful):

Btw in this use case, it is even irrelevant to provide the `blockProposer`, as AFAIK it is used only to determine who receives the award for proposing the block. So in this case, even `types.ZeroAddress` is ok.
```suggestion
    blockProposer := types.BytesToAddress(block.Header.Miner)
	transition, err := j.BeginTxn(parentHeader.StateRoot, block.Header, blockProposer)
```